File: projhub/cli/agent.py
from __future__ import annotations
from pathlib import Path
from typing import Optional

import typer
from ._console import console

from ..lib.config import find_project_root
from ..lib.markdown import parse_frontmatter

app = typer.Typer(help="Manage agent definitions")


def _require_root() -> Path:
    root = find_project_root()
    if not root:
        console.print("[red]No .projhub/ found. Run `projhub init` first.[/red]")
        raise typer.Exit(1)
    return root


@app.command("list")
def agent_list():
    """List configured agents."""
    root = _require_root()
    agents_dir = root / ".projhub" / "agents"
    if not agents_dir.exists():
        console.print("[dim]No agents configured.[/dim]")
        return

    for f in sorted(agents_dir.glob("*.md")):
        data, _ = parse_frontmatter(f.read_text(encoding="utf-8"))
        model = data.get("model", "standard")
        trigger = data.get("trigger", "ticket")
        model_color = "magenta" if model == "reasoning" else "dim" if model == "fast" else "cyan"
        name = data.get("name", f.stem)
        console.print(
            f"  [bold]{name:<16}[/bold] [{model_color}]{model:<10}[/{model_color}] "
            f"[dim]{trigger:<16}[/dim] {data.get('description', '')}"
        )


@app.command("add")
def agent_add(
    name: str = typer.Argument(..., help="Agent name"),
    from_template: Optional[str] = typer.Option(None, "--from", help="Base on an existing agent"),
):
    """Create a new agent definition."""
    root = _require_root()
    agents_dir = root / ".projhub" / "agents"
    target_path = agents_dir / f"{name}.md"

    if target_path.exists():
        console.print(f"[yellow]Agent {name} already exists.[/yellow]")
        raise typer.Exit(1)

    if from_template:
        source = agents_dir / f"{from_template}.md"
        if not source.exists():
            console.print(f"[red]Template agent {from_template} not found.[/red]")
            raise typer.Exit(1)
        import re
        content = source.read_text(encoding="utf-8")
        content = re.sub(r"^(name:\s*).*$", rf"\g<1>{name}", content, flags=re.MULTILINE)
        target_path.write_text(content, encoding="utf-8")
    else:
        target_path.write_text(
            f"""---
name: {name}
description: "(describe what this agent does)"
model: standard
tools: [read, search, edit, write, shell]
trigger: ticket
---

# Identity

You are the **{name}** agent. (Define identity and purpose)

# Guard Rail

(Define when this agent should refuse to work)

# Scope

(Define what this agent does and does NOT do)

# Work Order

1. (Step-by-step work process)

# Report Format

- **Done**: bullets with file:line references.
- **Definition of Done**: each Goal item marked `[x]` or `[ ]`.
- **Suggested next step**: 1 line.
""",
            encoding="utf-8",
        )

    console.print(f"[green]Created agent: .projhub/agents/{name}.md[/green]")

File: src/cli/commands/agent.js
import fs from 'node:fs';
import path from 'node:path';
import chalk from 'chalk';
import { findProjectRoot } from '../../lib/config.js';
import { parseFrontmatter } from '../../lib/markdown.js';

export function registerAgentCommand(program) {
  const agent = program
    .command('agent')
    .description('Manage agent definitions');

  agent
    .command('list')
    .alias('ls')
    .description('List configured agents')
    .action(() => {
      const root = findProjectRoot();
      if (!root) {
        console.error(chalk.red('No .holoctl/ found. Run `holoctl init` first.'));
        process.exit(1);
      }

      const agentsDir = path.join(root, '.holoctl', 'agents');
      if (!fs.existsSync(agentsDir)) {
        console.log(chalk.dim('No agents configured.'));
        return;
      }

      const files = fs.readdirSync(agentsDir).filter(f => f.endsWith('.md'));
      for (const file of files) {
        const content = fs.readFileSync(path.join(agentsDir, file), 'utf8');
        const { data } = parseFrontmatter(content);
        const model = data.model || 'standard';
        const trigger = data.trigger || 'ticket';
        const modelColor = model === 'reasoning' ? chalk.magenta : model === 'fast' ? chalk.dim : chalk.cyan;
        console.log(
          `  ${chalk.bold((data.name || file).padEnd(16))} ${modelColor(model.padEnd(10))} ${chalk.dim(trigger.padEnd(16))} ${data.description || ''}`
        );
      }
    });

  agent
    .command('add <name>')
    .description('Create a new agent definition')
    .option('--from <template>', 'Base on an existing agent template')
    .action((name, opts) => {
      const root = findProjectRoot();
      if (!root) {
        console.error(chalk.red('No .holoctl/ found. Run `holoctl init` first.'));
        process.exit(1);
      }

      const agentsDir = path.join(root, '.holoctl', 'agents');
      const targetPath = path.join(agentsDir, `${name}.md`);

      if (fs.existsSync(targetPath)) {
        console.error(chalk.yellow(`Agent ${name} already exists.`));
        process.exit(1);
      }

      if (opts.from) {
        const sourcePath = path.join(agentsDir, `${opts.from}.md`);
        if (!fs.existsSync(sourcePath)) {
          console.error(chalk.red(`Template agent ${opts.from} not found.`));
          process.exit(1);
        }
        let content = fs.readFileSync(sourcePath, 'utf8');
        content = content.replace(/^(name:\s*).*$/m, `$1${name}`);
        fs.writeFileSync(targetPath, content, 'utf8');
      } else {
        fs.writeFileSync(targetPath, `---
name: ${name}
description: "(describe what this agent does)"
model: standard
tools: [read, search, edit, write, shell]
trigger: ticket
---

# Identity

You are the **${name}** agent. (Define identity and purpose)

# Guard Rail

(Define when this agent should refuse to work)

# Scope

(Define what this agent does and does NOT do)

# Work Order

1. (Step-by-step work process)

# Report Format

- **Done**: bullets with file:line references.
- **Definition of Done**: each Goal item marked \`[x]\` or \`[ ]\`.
- **Suggested next step**: 1 line.
`, 'utf8');
      }

      console.log(chalk.green(`Created agent: .holoctl/agents/${name}.md`));
    });
}
